### ISSUE TITLE
Repeated ID_name/AllWISE/ObjID values

### ISSUE THREAD

**Issue by bmingo** [mhardcastle/lgz#20]
There seem to be some catalogue entries in version 0.6 (also present in version 0.5) in which the same ID_name, AllWISE, and/or ObjID value is assigned to different Source_Name values (with different fluxes and slightly different coordinates).

In the case of the repeated ID_name, there is at least one known issue covered (entries with ID_name="mult" on the catalogue), but I seem to have picked up a further 49 pairs of sources with identical ID_names, all of which also have identical AllWISE and ObjID values (when those are present). All the pairs seem to also have had the IDs assigned via the same method (same ID_flag).

There are 97 pairs of sources with the same AllWISE value (partially overlapping the previous group). Some of these have a different ID_name for each source, and different values of ID_flag. Where available, all pairs also have the same ObjID value.

There are 82 pairs of sources with the same ObjID value. This group seems to be a subset of the previous one, as in all cases the AllWISE values are identical for both sources in each pair (there are two pairs where allWISE=N/A, and ID_name and ID_flag have different values, so this selection by ObjID is still necessary to catch those cases).

I investigated this issue after finding that this galaxy zoo source had two separate entries in the catalogue (one matched through galaxy zoo, the other matched to a bright galaxy), which used different combinations of components:
https://www.zooniverse.org/projects/mjh22/lofar-galaxy-zoo/talk/subjects/10509209

I believe there may be two possible issues here: 
The first is where one of the matches to the same optical source has been achieved through galaxy zoo, and the second through maximum likelihood or match to a bright galaxy.
The second case, more prevalent for repeated values of ID_name, has the same ID_flag for both entries in the pair with the same optical ID, likely meaning that different combinations of components are used for the same source through the same method (galaxy zoo or maximum likelihood).

I have saved the tables for the three sets of duplicate entries on the Herts LOFAR server, on
/data/lofar/bmingo/morph/hetdex/tests

**Comment by mhardcastle** [comment]
For the LGZ ones (id 311) a lot of these are the result of large artefacts being centred on a smaller bright source: both are identified with the host without being associated with each other. The artefacts will just need to be removed manually. I'll set up a mechanism for this.

**Comment by mhardcastle** [comment]
The ones with ID_flag==1 are basically all real sources where two concentric components have not been associated by PyBDSF. These should really be a separate flowchart branch but once located they're easy to fix by sending them through the 'too zoomed in' code, which I have now done.

**Comment by mhardcastle** [comment]
@bmingo Can you have a look at today's catalogue release v0.7 and tell us whether there are still problems?

**Comment by bmingo** [comment]
I'll have a look first thing tomorrow!

**Comment by bmingo** [comment]
Sorry about the delay.

There are still some issues, it seems:

71 pairs of sources have the same AllWISE value. All of them seem to have the same ObjID values, where present (5 pairs have empty ObjIDs). In most cases, both IDs have been achieved through different methods (different ID_flag values, also different ID_name), but there are 3 cases with identical ID_flags (2x602, 1x311) and ID_names.

There are 71 groups of sources with duplicate (in one case, triplicate) ObjID values. This group overlaps almost completely with the previous one, except for 5 sets (including the triplicate) where AllWISE="N/A". For these 5 sets the ID_flag and ID_name for each member of the pair is different, except for the triplicate, where ID_flag=311 for 2/3 sources, which also have the same ID_name (the third source has an almost identical ID name, with 2MASX J1... instead of 2MASS 1...).

There are 4 pairs of sources with duplicate ID_Name values (other than the ones with ID_Name=Mult). This is a subset of the previous 2 groups (3 groups have duplicate AllWISE and ObjID values, the fourth one has AllWISE=N/A, and duplicated ObjID).

I have saved these tables on the Herts LOFAR server, in /data/lofar/bmingo/tests, for your perusal.

**Comment by wndywllms** [comment]
Just looked at one or 2 examples... some of these look like this 
ILTJ150057.24+461721.9:
![image](https://user-images.githubusercontent.com/2170761/35569853-4896ccb8-05c5-11e8-8d4f-b43152681f34.png)
where either one or both of the sources has been in the LGZ sample (I think there is a similar case where they both come directly from ML only matching). 

In the LGZ samples: in the case where only one source was in LGZ it would be larger one, which should either have been associated with the small one or marked as an artefact. Can it be that the association is not picked up even though users have clicked on the small source because the click was nearer to the centre of the large source?

**Comment by bmingo** [comment]
@jcroston might be able to check that...

**Comment by mhardcastle** [comment]
I believe the code is meant to ignore the source with the cross in processing clicks, so a click on the small source should go through. I think it's more likely that people didn't notice the second component in some cases.

In any case we can deal with these by manually fixing them up with the too-zoomed-in pipeline, hopefully -- which I thought I had already done for most of them, but perhaps I missed them. 602s are more of a problem since these should always map on to distinct optical sources. I wonder if there's an issue with mismatches between the ML catalogue and Ken's catalogue? The final match with the latter is just done by position...

**Comment by mhardcastle** [comment]
So, just to be clear, ID_name is assigned by multiple routes dependent on ID_flag. There's no expectation that sources with the same optical ID will have the same ID_name at the moment (this will be fixed eventually, it's issue #6). On the other hand ID_ra, ID_dec and all the columns from AllWISE on are from the crossmatch with Ken's photo-z catalogue and so there should be no disagreement between them (i.e. no cases where AllWISE is the same but PS ID is different or whatever).

What we can't see from this is what the RA and DEC for the optical position was _before_ the match, since the code assimilates these with what it finds in the master optical catalogue.

**Comment by bmingo** [comment]
How does the matching process work for ID_flag=2? I have had a look at a couple of cases where the pairs have ID_flag=1/2, and they seem to have similar configurations to the one posted by Wendy (where the component with ID_flag=1 is internal to the one with ID_flag=2, and the latter has LGZ information).
See e.g.
ILTJ105900.17+500254.8
ILTJ105900.56+500259.5

**Comment by wndywllms** [comment]
Ok, I think I fixed the ID_flag 1/2 combination with e8d3679eea019af05dd7c361a345d58abdbd52c3 (not quite the bug I thought it was).

ID_flag 2 comes from 2 routes: either automatically merge everything that matches a large (60") 2MASX galaxy, or it's a large/faint radio source that a visual check led to 'match to the bright optical source' (in most cases to the recorded 2MASX match). In the latter case there are a few instances where there is a small radio source which has a match to the same 2MASX source which I missed in merging.

**Comment by wndywllms** [comment]
In the latest version (v0.8) there are still 374 sources that share the same 180 ID_name's

Most of these look like there is a small source that has been missed in the LGZ association (maybe making them cyan because they have an ML match made them less visible than before)

There is a text file listing Source_Name ID_flag Assoc Component_Names Assoc_Qual for each group of sources with the same ID.
[double_ID_list.txt](https://github.com/mhardcastle/lotss-catalogue/files/1702928/double_ID_list.txt)

One weird one is where 4 sources (each of which was in LGZ, so really should have increased the click statistics on this source...) have not been associated but each has the ID of what should be the association:
Group  13 : ID= PSO J110938.811+494609.490 
ILTJ110938.07+494556.6 311 1 ['ILTJ110938.07+494556.6'] nan
ILTJ110938.39+494602.8 311 1 ['ILTJ110938.39+494602.8'] nan
ILTJ110940.26+494626.6 311 1 ['ILTJ110940.26+494626.6'] nan
ILTJ110940.52+494630.0 311 1 ['ILTJ110940.52+494630.0'] nan

**Comment by mhardcastle** [comment]
Is there any reason why the numbering of the groups starts at 2?

**Comment by wndywllms** [comment]
yes -- Group 1 is the group of ~3600 ID_name=='mult' sources

**Comment by mhardcastle** [comment]
Ah right. OK. I am going to feed all the ones where we have a pair of a LGZ and ML source through the too-zoomed-in pipeline, then, which should reduce these to a smaller number. Just wanted to check that the file was complete before starting to make a source list.

**Comment by mhardcastle** [comment]
This should be mostly done in v0.8, but we need a separate piece of code to deal with the blends, so I won't close it yet.

**Comment by wndywllms** [comment]
> The ones with ID_flag==1 are basically all real sources where two concentric components have not been associated by PyBDSF. These should really be a separate flowchart branch but once located they're easy to fix by sending them through the 'too zoomed in' code, which I have now done.

@mhardcastle I think we picked up a few more of these after the LRs changed, can you run them through the zoom code as before?

Source_Name         Total_flux             ID_name           ID_flag     ML_LR    

ILTJ112844.18+540033.4 0.784852894044 AllWISE J112844.70+540031.8       1 4.73230873371
ILTJ112844.65+540031.3  6.31202608793 AllWISE J112844.70+540031.8       1 707.949027643

ILTJ115355.18+523852.2 2.24303873738 PSO J115355.978+523845.763       1 25.3793026662
ILTJ115355.27+523855.0 4.43724528094 PSO J115355.978+523845.763       1 4.13245346328

ILTJ130951.78+460555.1 1.03025939679 PSO J130951.859+460600.366       1 3.52903080406
ILTJ130951.86+460559.6 9.90151903184 PSO J130951.859+460600.366       1 49.2675668778

ILTJ132642.93+460752.2 0.668015573022 PSO J132644.109+460748.728       1 4.27609692302
ILTJ132644.17+460749.6  2.07534553011 PSO J132644.109+460748.728       1 370.063268473

ILTJ134414.45+513510.3 0.510709847001 PSO J134414.751+513459.632       1 5.11947920317
ILTJ134414.73+513459.7  1.90469191967 PSO J134414.751+513459.632       1 3.71924444234

**Comment by mhardcastle** [comment]
These are all done now.

On the other hand I stumbled over some ID_flag==2 sources which seem only to have a partial match to the underlying bright galaxy, which means that two sources in the catalogue then have the same ID (or close to the same -- obviously the ID matches to Ken's catalogue may well be pretty meaningless in this case). These include (not an exhaustive list):
```
ILTJ113648.25+541733.5
ILTJ113324.32+470246.4
ILTJ133010.20+471614.2
ILTJ133005.80+470946.4
ILTJ121902.18+471349.9
ILTJ120624.58+524209.0
ILTJ140314.73+542314.7
```
You can see example images of these ones in `/data/lofar/mjh/hetdex_4/opt_bright/`.

What I think is happening here is that LGZ results are superseding the automatic crossmatching with galaxies for some parts of these (sometimes LGZ is getting it right, more often it is getting it wrong). We probably need to figure out how to fix these. Some are just a matter of putting in an override file to fix up the LGZ (and I have already done that in a few cases) but others (e.g. ILTJ140314.73+542314.7) are more difficult because what's left is a collection of point sources that LGZ decided don't belong with the original galaxy.

We should at least be able to catch this situation at the point where the LGZ catalogue merges with the flowchart one...

**Comment by mhardcastle** [comment]
The ID_flag==2 problem is fixed but there are still some sources with duplicate optical IDs in the v1.1 draft catalogue. This is now a test case, see #39

**Comment by mhardcastle** [comment]
Fixed in the released version 1.1.
